db/dcrpg/pgblockchain.go

```diff
@@ -3259,83 +3260,147 @@ type MsgBlockPG struct {
-	dbAddressRows := make([][]dbtypes.AddressRow, len(dbTransactions))
-	var totalAddressRows int
+	var voutStmt *sql.Stmt
+	voutStmt, err = dbTx.Prepare(internal.MakeVoutInsertStatement(checked, doUpsert))
+	if err != nil {
+		_ = dbTx.Rollback()
+		err = fmt.Errorf("failed to prepare vout insert statment: %v", err)
+		return
+	}
+	defer voutStmt.Close()
 
-	// For a side chain block, set Validators to an empty slice so that there
-	// will be no misses even if there are less than 5 votes. Any Validators
-	// that do not match a spent ticket hash in InsertVotes are considered
-	// misses. By listing no required validators, there are no misses. For side
-	// chain blocks, this is acceptable and necessary because the misses table
-	// does not record the block hash or main/side chain status.
-	if !isMainchain {
-		msgBlock.Validators = []string{}
+	var vinStmt *sql.Stmt
+	vinStmt, err = dbTx.Prepare(internal.MakeVinInsertStatement(checked, doUpsert))
+	if err != nil {
+		_ = dbTx.Rollback()
+		err = fmt.Errorf("failed to prepare vin insert statment: %v", err)
+		return
 	}
+	defer vinStmt.Close()
 
-	var err error
-	for it, dbtx := range dbTransactions {
+	// dbAddressRows contains the data added to the address table, arranged as
+	// [tx_i][addr_j], transactions paying to different numbers of addresses.
+	dbAddressRows = make([][]dbtypes.AddressRow, len(txns))
+
+	for it, Tx := range txns {
 		// Insert vouts, and collect AddressRows to add to address table for
 		// each output.
-		dbtx.VoutDbIds, dbAddressRows[it], err = InsertVouts(pgb.db,
-			dbTxVouts[it], pgb.dupChecks, updateExistingRecords)
+		Tx.VoutDbIds, dbAddressRows[it], err = InsertVoutsStmt(voutStmt,
+			vouts[it], pgb.dupChecks, updateExistingRecords)
 		if err != nil && err != sql.ErrNoRows {
-			log.Error("InsertVouts:", err)
-			txRes.err = err
-			return txRes
+			err = fmt.Errorf("failure in InsertVoutsStmt: %v", err)
+			_ = dbTx.Rollback()
+			return
 		}
 		totalAddressRows += len(dbAddressRows[it])
-		txRes.numVouts += int64(len(dbtx.VoutDbIds))
-		if err == sql.ErrNoRows || len(dbTxVouts[it]) != len(dbtx.VoutDbIds) {
+		numOuts += len(Tx.VoutDbIds)
+		if err == sql.ErrNoRows || len(vouts[it]) != len(Tx.VoutDbIds) {
 			log.Warnf("Incomplete Vout insert.")
 		}
 
 		// Insert vins
-		dbtx.VinDbIds, err = InsertVins(pgb.db, dbTxVins[it], pgb.dupChecks,
+		Tx.VinDbIds, err = InsertVinsStmt(vinStmt, vins[it], pgb.dupChecks,
 			updateExistingRecords)
 		if err != nil && err != sql.ErrNoRows {
-			log.Error("InsertVins:", err)
-			txRes.err = err
-			return txRes
+			err = fmt.Errorf("failure in InsertVinsStmt: %v", err)
+			_ = dbTx.Rollback()
+			return
 		}
-		txRes.numVins += int64(len(dbtx.VinDbIds))
+		numIns += len(Tx.VinDbIds)
 
 		// return the transactions vout slice if processing stake tree
-		if txTree == wire.TxTreeStake {
-			dbtx.Vouts = dbTxVouts[it]
-		}
+		//if txTree == wire.TxTreeStake {
+		Tx.Vouts = vouts[it]
+		//}
```

Comment:
Is this block of code helpful for the regular Tx Tree type?  I can only see `Tx.Vouts` contents only being used in `InsertTickets` function

Reply:
You are right that it is only used in `InsertTickets`.  But I've removed `txTree` from the input arguments to `storeTxns` because I was uneasy about having the `Vouts` set only for stake transactions.  It would perhaps be a surprise to find it empty for the regular transaction tree if you wanted it.  Since it's copying slices, rather than the data in the slices, it's not very costly to do.